src/components/FullHierarchyTable/FullHierarchyTable.tsx

```diff
@@ -175,16 +318,66 @@ export const FullHierarchyTable: React.FunctionComponent<FullHierarchyTableAllPr
         }
     }
 
+    const customDropdown = (dropdownProps: FilterDropdownProps, key: string) =>
+        <div className={styles.filterContent}>
+            <Form layout="vertical">
+                <Input
+                    autoFocus
+                    placeholder={inputPlaceholder}
+                    value={dropdownProps.selectedKeys[0]}
+                    suffix={<Icon type="search"/>}
+                    onChange={(e) => {
+                        dropdownProps.setSelectedKeys(e.target.value ? [e.target.value] : [])
+                    }}
+                    maxLength={50}
+                />
+                <div className={styles.operators}>
+                    <Button className={styles.button} htmlType="submit" onClick={() => handleApply(
+                        dropdownProps.confirm,
+                        dropdownProps.selectedKeys,
+                        key)}>
+                        {t('Apply')}
+                    </Button>
+                    <Button className={styles.button} onClick={() => handleCancel(
+                        dropdownProps.clearFilters,
+                        dropdownProps.selectedKeys,
+                        key)}>
+                        {t('Clear')}
+                    </Button>
+                </div>
+            </Form>
+        </div>
+
+    const dropDown = (filterableFieldsKey: RowMetaField[], key: string) => {
```

Comment:
Thanks for the code review and comments, they are relevant. The component was really confusing, the intention to simplify it is obvious.
I followed the steps listed and got the result. However, I have comments:

**Step 2** `props.filters?.find`
```ts
const textFilters = React.useMemo(() => props.filters?.find(filter =>
        [FilterTypes.contains, FilterTypes.equals].includes(filter.type),
    [props.filters]
)
```

Now `props.bcFilter` replace `textFilters`


**Step 5** 
```ts
bcFilterMatchedAncestors(item, props.data, props.bcFilter)?.forEach(key => result.push(key))`
Type mismatch, explicit casting is needed:
`bcFilterMatchedAncestors(item as FullHierarchyDataItem, props.data as FullHierarchyDataItem[], textFilters)
```


The proposed data filter algorithm does not take into account the fact that both ancestors and descendants of data elements that fall under the search conditions must be present in the result set. This is an important point.


I divided the key constants of the ancestors and descendants so that they could be used separately (when calculating expand).

```ts
const searchedAncestorsKeys: Set<string> = React.useMemo(() => {
        const result: string[] = []
        props.data.forEach(item => {
            bcFilterMatchedAncestors(item as FullHierarchyDataItem, props.data as FullHierarchyDataItem[], textFilters)
                ?.forEach(key => result.push(key))
        })
        return new Set(result)
    }, [props.data, textFilters])
```

```ts
const searchedDescendantsKeys: Set<string> = React.useMemo(() => {
        const result: string[] = []
        props.data.forEach(item => {
            bcFilterMatchedDescendants(item as FullHierarchyDataItem, props.data as FullHierarchyDataItem[], textFilters)
                ?.forEach(key => result.push(key))
        })
        return new Set(result)
    }, [props.data, textFilters])
```
	
	
and got the data like that:

```ts
const filteredData = React.useMemo(() => {
        return textFilters?.length
            ? props.data.filter(item => searchedAncestorsKeys.has(item.id) || searchedDescendantsKeys.has(item.id))
            : props.data
    }, [searchedAncestorsKeys, searchedDescendantsKeys, props.data, textFilters])
```
	
	
Given the fact of the presence of descendants in the data tree, the proposed algorithm will be changed. It is necessary to take into account that now the required elements can have descendants, but they do not need to be expanded.


Get three subsets of elements and collect them into one:
- manual expanded keys
- expanded keys of associated elements
- filtered expanded keys as all ancestors of filtered elements without last children of subtree


This set of changes leads to expected result and has correctly structured readable code.

Reply:
component modified using existing `<ColumnTitle>`